### PR TITLE
Assignment5 - Josh Pedersen & Conner Hanson

### PR DESCRIPTION
**Pull request by Joshua-Pedersen**
Partner: @connerhanson

### CODE REVIEW

**Comment by SergeiAshikhmin** [comment]
I will review this

**Review by SergeiAshikhmin — COMMENTED** [review]
Code looks good and clean. Good job 👍

Instructions
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
The values should be homogeneous.
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
Create an Exists method to test to see if a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔

I did not see any mentions of extra credit so I assume you skipped it.

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Comment by Matthew-Powers** [comment]
I will review this

File: GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        // To add once Exists method is here
+        if (Exists(value)) 
+        {
+            throw new System.ArgumentException("Duplicate Value");
+        }
+         
+
+        Node<T> newNode = new(value);
+        Node<T> cur = this;
```
Comment (Matthew-Powers) [inline GenericsHomework/Node.cs:30]: cur isn't a very descriptive variable name. I'm assuming it's short for 'current'. If so, replace cur with current

File: GenericsHomework/Node.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    public override string ToString()
+    {
+            string contents = $"{Value}";
+            return contents;
+    }
+
+    public void Append(T value)
+    {
+        // To add once Exists method is here
```
Comment (Matthew-Powers) [inline GenericsHomework/Node.cs:22]: Exists method is here, don't need this comment any more

File: GenericsHomework/Node.cs (hunk 1 above)
Comment (Matthew-Powers) [reply GenericsHomework/Node.cs:30]: Even better might be currentNode

File: GenericsHomework/Node.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        if (Exists(value)) 
+        {
+            throw new System.ArgumentException("Duplicate Value");
+        }
+         
+
+        Node<T> newNode = new(value);
+        Node<T> cur = this;
+
+        while (cur.Next != this) { cur = cur.Next; }
```
Comment (Matthew-Powers) [inline GenericsHomework/Node.cs:32]: This probably should be broken up into multiple lines for readability
```suggestion
        while (cur.Next != this) 
        { 
            cur = cur.Next; 
        }

```

File: GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        // base case if list is a single element
+        if(this.Next == this)
+        {
+            return;
+        }
+
+        // Reference of next node before clearing 
+        Node<T> curNode = this;
+
+        // Loop through list until your at the node before this
```
Comment (Matthew-Powers) [inline GenericsHomework/Node.cs:49]: nitpick: you're*

File: GenericsHomework/Node.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    public Node(T value)
+    {
+        Value = value;
+        Next = this;
+    }
+
+    public override string ToString()
+    {
+            string contents = $"{Value}";
```
Comment (Matthew-Powers) [inline GenericsHomework/Node.cs:16]: Kind of weird spacing going on here. Should be reduced to 1 tab

File: GenericsHomework/Node.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
+﻿namespace GenericsHomework;
+
+public class Node<T>
```
Comment (Matthew-Powers) [inline GenericsHomework/Node.cs:3]:
```suggestion
public class Node<T> : System.Collections.Generic.ICollection<T>

```

File: Directory.Build.Props [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿<Project>
+	<PropertyGroup>
+		<ImplicitUsings>enable</ImplicitUsings>
+		<LangVersion>12.0</LangVersion>
+		<Nullable>enable</Nullable>
```
Comment (Matthew-Powers) [inline Directory.Build.Props:5]: <TargetFramework>net8.0</TargetFramework>

Comment (Matthew-Powers) [inline Directory.Build.Props:5]: Missing \<TargetFramework>net8.0\</TargetFramework>

File: GenericsHomework.Tests/GenericsHomework.Tests.csproj [hunk 8]
```diff
@@ -0,0 +1,26 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+	<PropertyGroup>
+		<TargetFramework>net8.0</TargetFramework>
```
Comment (Matthew-Powers) [inline GenericsHomework.Tests/GenericsHomework.Tests.csproj:4]: Can remove this after putting it in the props file

File: GenericsHomework.Tests/NodeTests.cs [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    {
+        // Arrange
+        Node<int> node = new(1);
+        node.Append(2);
+        node.Append(3);
+
+
+        // Act and Assert
+        Assert.Throws<System.ArgumentException>(() => node.Append(1));
+        Assert.Throws<System.ArgumentException>(() => node.Append(3));
```
Comment (Matthew-Powers) [inline GenericsHomework.Tests/NodeTests.cs:95]: Should probably only use a single assert

File: GenericsHomework/GenericsHomework.csproj [hunk 10]
```diff
@@ -0,0 +1,10 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+	<PropertyGroup>
+		<TargetFramework>net8.0</TargetFramework>
```
Comment (Matthew-Powers) [inline GenericsHomework/GenericsHomework.csproj:4]: Can remove this after adding it to props file

File: GenericsHomework/Node.cs (hunk 6 above)
Comment (Matthew-Powers) [reply GenericsHomework/Node.cs:3]: For extra credit

**Review by Matthew-Powers — COMMENTED** [review]
Instructions

    Create a class library project called "GenericsHomework.". ✔
    Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item. ❌
--Nodes don't point back to the first item (might not actually be an issue, since this is kind of vague in terms of meaning)
        Define the Node class ✔
        The values should be homogeneous. ✔
        Include a constructor that takes a value. (No validation is necessary on the value). ✔
        Add a ToString() override that writes out the value's ToString() result. ✔
        Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
            The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
            The Next property setter should be private. ✔
        Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
        Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
            Whether it is sufficient to only set Next to itself ✔
            Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
            Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
        Create an Exists method to test to see if a value exists in the list. ✔
        Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
    You should not rely on any BCL generic classes for your implementation. ✔

Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

    Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌

    Each circle contains n items and each item can belong to one or more Circle instances. ❌
    You are not required to use a Node from earlier in the homework for your Venn diagram implementation. ❌
    You are welcome to use existing BCL generic classes for the extra credit. ❌

    Implement System.Collections.Generic.ICollection<T> on the Node class ❌

Fundamentals

    Place all shared project properties into a Directory.Build.Props file. ❌
--Missing: <TargetFramework>net8.0</TargetFramework>in props file
    Place all shared project items into a Directory.Build.targets file. ❌
--File does not exist
    nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ✔
    For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
--Never used in test, and I see a lot of Assert.Equal
    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔

Overall looks pretty good. Just needs some tweaking.

Would also recommend looking into cleaning up your whitespace. EG: Line 18 in NodeTests

File: GenericsHomework/Node.cs (hunk 1 above)
Comment (connerhanson) [reply GenericsHomework/Node.cs]: Changed all instances of cur or curNode to currentNode.  Naming it cur is just a habit I got from doing linked lists in for data structures, thanks for the advice!

File: GenericsHomework/Node.cs (hunk 2 above)
Comment (connerhanson) [reply GenericsHomework/Node.cs]: Fixed, thanks for the reminder!

File: GenericsHomework/Node.cs (hunk 3 above)
Comment (connerhanson) [reply GenericsHomework/Node.cs]: Fixed, Good suggestion!

File: GenericsHomework.Tests/NodeTests.cs (hunk 9 above)
Comment (connerhanson) [reply GenericsHomework.Tests/NodeTests.cs:95]: Here I used two asserts to make sure the exist method loops properly.  First making sure it finds the first node, and second making sure it finds the last node.

**Comment by connerhanson** [comment]
@Matthew-Powers  For this assignment, always use Assert.AreEqual() (the generic version) ❌ --Never used in test, and I see a lot of Assert.Equal 

In xunit Assert.Equal is generic by default.  Found in this document: https://xunit.net/docs/comparisons

File: GenericsHomework/Node.cs (hunk 4 above)
Comment (Joshua-Pedersen) [reply GenericsHomework/Node.cs:51]: Good catch lol

File: GenericsHomework/Node.cs (hunk 6 above)
Comment (Joshua-Pedersen) [reply GenericsHomework/Node.cs:3]: Implementation of ICollection will make the code messy, we're choosing to stick to the base tasks.

File: Directory.Build.Props (hunk 7 above)
Comment (Joshua-Pedersen) [reply Directory.Build.Props:5]: We moved the ``<TargetFramework>`` into each of the csproj files instead. Placing the ``<TargetFramework>`` in the ``.props`` file causes a fatal error where neither the BuildTest nor certain operating systems are able to build the code.

File: GenericsHomework/GenericsHomework.csproj (hunk 10 above)
Comment (Joshua-Pedersen) [reply GenericsHomework/GenericsHomework.csproj:4]: We moved the <TargetFramework> into each of the csproj files instead. Placing the <TargetFramework> in the .props file causes a fatal error where neither the BuildTest nor certain operating systems are able to build the code.

File: GenericsHomework.Tests/NodeTests.cs [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    public void ToString_HasValue_ReturnsString()
+    {
+        // Arrange
+        int val =  42;
+
+        // Act
+        Node<int> testNode = new(val);
+
+        // Assert
+        Assert.Equal($"{val}", testNode.ToString());
```
Comment (BenjaminMichaelis) [inline GenericsHomework.Tests/NodeTests.cs:31]:
```suggestion
        Assert.Equal(val.ToString(), testNode.ToString());

```
Why not `.ToString()`?

File: GenericsHomework/GenericsHomework.csproj [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+	</PropertyGroup>
+	<ItemGroup>
+		<None Remove="Directory.Build.Props"/>
+	</ItemGroup>
```
Comment (BenjaminMichaelis) [inline GenericsHomework/GenericsHomework.csproj:8]: What's this for?

File: GenericsHomework/Node.cs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        Value = value;
+        Next = this;
+    }
+
+    public override string ToString()
+    {
+        string contents = $"{Value}";
+        return contents;
+    }
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:18]:
```suggestion
    public override string? ToString()
    {
        if (Value == null) return null;
        return Value.ToString();
    }

```

slightly improved

File: GenericsHomework/Node.cs [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        Node<T> newNode = new(value);
+        Node<T> currentNode = this;
+
+        while (currentNode.Next != this) 
+        { 
+            currentNode = currentNode.Next; 
+        }
+
+        currentNode.Next = newNode;
+        newNode.Next = this;
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:37]:
```suggestion
         
        Node<T> newNode = new(value)
        {
            Next = Next
        };
        Next = newNode;
        return newNode;
```
bit simpler logic

File: GenericsHomework/Node.cs [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+        while (currentNode.Next != this)
+        {
+            currentNode = currentNode.Next;
+        }
+
+        // Set node before this to this's next node
+        currentNode.Next = this.Next;
+
+        // Clear by setting the node to itself 
+        this.Next = this;
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:61]: I think all you need for this method is this line

File: GenericsHomework/Node.cs [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    public bool Exists(T value)
+    {
+        // create node to navigate with starting with the first node
+        Node<T> currentNode = this;
+
+        // since curNode cannot equal this/first node in the loop, we need to do an 
+        // intial test to make sure the first node does not have the given value.
+
+        // Use of null forgiveness since doc states validation on node values is not nessasary
+        if(currentNode.Value!.Equals(value))
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:80]: What if T is nullable and has a null value? Shouldn't it return true if both Value and value are null?

File: GenericsHomework/Node.cs [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+        // loop until initial node, if values are equal return true, otherwise return false.
+        while(currentNode != this)
+        {
+            // Use of null forgiveness since doc states validation on node values is not nessasary
+            if (currentNode.Value!.Equals(value))
+            {
+                return true;
+            }
+            currentNode = currentNode.Next;
```
Comment (BenjaminMichaelis) [inline GenericsHomework/Node.cs:96]: A do while loop can reduce this complication

**Pull request closed without merging**